producer/buffer/buffer.go

```diff
@@ -116,52 +129,50 @@ func (b *buffer) Add(m producer.Message) (*producer.RefCountedMessage, error) {
 		b.m.messageTooLarge.Inc(1)
 		return nil, errMessageTooLarge
 	}
-	b.Lock()
+	b.RLock()
 	if b.isClosed {
-		b.Unlock()
+		b.RUnlock()
 		return nil, errBufferClosed
 	}
-	dataSize := uint64(s)
-	targetBufferSize := b.maxBufferSize - dataSize
-	if b.size.Load() > targetBufferSize {
-		if err := b.produceOnFullWithLock(targetBufferSize); err != nil {
-			b.Unlock()
+	messageSize := uint64(s)
+	newBufferSize := b.size.Add(messageSize)
+	if newBufferSize > b.maxBufferSize {
+		if err := b.produceOnFull(newBufferSize, messageSize); err != nil {
+			b.RUnlock()
 			return nil, err
 		}
 	}
-	b.size.Add(dataSize)
 	rm := producer.NewRefCountedMessage(m, b.onFinalizeFn)
-	b.buffers.PushBack(rm)
-	b.Unlock()
+	b.listLock.Lock()
```

Comment:
Do we need two locks? Keep in mind that with two locks you need to ensure you don't accidentally introduce deadlocks, which is tricky especially given that the code here is already fairly complex so I'd suggest using one lock if at all possible.

Reply:
The main contention is on the listLock, which is also what I'm trying to reduce here, the usage of it is always paired with `Lock` and `Unlock` closely, so it's cleaner, the embedded RWMutex is just guarding some variables like `isClosed`, etc

Reply:
Okay please make sure you don't lock them in different orders while holding one of them.